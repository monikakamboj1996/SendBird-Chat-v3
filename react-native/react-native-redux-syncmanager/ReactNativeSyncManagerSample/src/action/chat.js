const Action = {
  Type: {
    MESSAGE_SELECT: 'message-select',
    MESSAGE_DESELECT_ALL: 'message-deselect-all',
    MESSAGE_ADD: 'message-add',
    MESSAGE_UPDATE: 'message-update',
    MESSAGE_REMOVE: 'message-remove',
    MESSAGE_CLEAR: 'message-clear'
  }
};
Action.selectMessage = message => {
  return {
    type: Action.Type.MESSAGE_SELECT,
    payload: {
      message
    }
  };
};
Action.deselectAllMessages = () => {
  return {
    type: Action.Type.MESSAGE_DESELECT_ALL
  };
};
Action.addMessages = messages => {
  return {
    type: Action.Type.MESSAGE_ADD,
    payload: {
      messages
    }
  };
};
Action.updateMessages = (messages, isTransformFailedToManual) => {
  return {
    type: Action.Type.MESSAGE_UPDATE,
    payload: {
      messages,
      isTransformFailedToManual
    }
  };
};
Action.removeMessages = messages => {
  return {
    type: Action.Type.MESSAGE_REMOVE,
    payload: {
      messages
    }
  };
};
Action.clearMessages = () => {
  return {
    type: Action.Type.MESSAGE_CLEAR
  };
};

export default Action;
