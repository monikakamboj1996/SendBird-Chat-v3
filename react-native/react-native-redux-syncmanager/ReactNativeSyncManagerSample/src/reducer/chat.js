import Action from '../action/chat';
import { findMessageIndex, getMessageIndex } from '../utils';

const _initialState = {
  channel: null,
  messages: [],
  inputMode: 'chat'
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.MESSAGE_SELECT: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = newState.messages[i].isIdentical(action.payload.message);
      }
      break;
    }
    case Action.Type.MESSAGE_DESELECT_ALL: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = false;
      }
      break;
    }
    case Action.Type.MESSAGE_ADD: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const isFailed = message.messageId === 0 && message.requestState === 'failed';
        const index = findMessageIndex(message, newState.messages, isFailed);
        if (index >= 0) {
          newState.messages.splice(index, 0, message);
          if (index > 0) {
            newState.messages[index - 1]._isPreviousMessageSentBySameUser =
              newState.messages[index - 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index - 1].sender.userId === newState.messages[index].sender.userId;
          }
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index].sender &&
              newState.messages[index + 1].sender &&
              newState.messages[index].sender.userId === newState.messages[index + 1].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_UPDATE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          message._isPreviousMessageSentBySameUser = newState.messages[index]._isPreviousMessageSentBySameUser;
          newState.messages[index] = message;
          if (action.payload.isTransformFailedToManual) {
            newState.messages[index]._isManualResend = true;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_REMOVE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          newState.messages.splice(index, 1);
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index + 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index + 1].sender.userId === newState.messages[index].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_CLEAR: {
      newState.messages = [];
      break;
    }
  }
  return newState;
};
