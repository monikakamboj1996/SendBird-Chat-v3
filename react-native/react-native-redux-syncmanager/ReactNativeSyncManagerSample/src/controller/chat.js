import React from 'react';
import { Alert, BackHandler, Clipboard, Keyboard, View } from 'react-native';
import { Button, Input } from 'react-native-elements';
import { connect } from 'react-redux';

import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';
import Permissions from 'react-native-permissions';
import ImagePicker from 'react-native-image-picker';

import SendBird from 'sendbird';
import SendBirdSyncManager from 'sendbird-syncmanager';

import { style } from '../style/chat';
import Action from '../action/chat';
import MessageView from '../view/message';
import { createChannelTitle, toast } from '../utils';
import { navigator } from '../navigator';

const PAGE_RETURN_DELAY = 600;
let componentContext = null;

const _chatCloseHeader = navigation => {
  return (
    <View style={style.headerLeftContainer}>
      <Button
        containerStyle={style.headerLeftItemContainer}
        buttonStyle={style.headerLeftItem}
        icon={<Icon name="md-arrow-back" color="#333" size={28} />}
        onPress={() => {
          navigation.goBack();
        }}
      />
    </View>
  );
};
const _chatControlHeader = navigation => {
  const channel = navigation.getParam('channel', null);
  return (
    <View style={style.headerRightContainer}>
      <Button
        containerStyle={style.headerRightItemContainer}
        buttonStyle={style.headerRightItem}
        icon={<Icon name="md-contacts" color="#7d62d9" size={28} />}
        onPress={() => {
          const sb = SendBird.getInstance();
          sb.GroupChannel.getChannel(channel.url, (channel, err) => {
            if (!err) {
              navigation.push('Member', { channel });
            } else {
              toast('Failed to get channel information.');
            }
          });
        }}
      />
      <Button
        containerStyle={style.headerRightItemContainer}
        buttonStyle={style.headerRightItem}
        icon={<Icon name="md-exit" color="#7d62d9" size={28} />}
        onPress={() => {
          Alert.alert('Leave', 'Do you want to leave this channel?', [
            { text: 'Cancel' },
            {
              text: 'OK',
              onPress: () => {
                channel.leave((res, err) => {
                  if (!err) {
                    navigation.goBack();
                  } else {
                    toast('Failed to leave channel.');
                  }
                });
              }
            }
          ]);
        }}
      />
    </View>
  );
};
const _menuCloseHeader = navigation => {
  const currentChat = navigation.getParam('currentChat', null);
  return (
    <View style={style.headerLeftContainer}>
      <Button
        containerStyle={style.headerLeftItemContainer}
        buttonStyle={style.headerLeftItem}
        icon={<Icon name="md-close" color="#333" size={28} />}
        onPress={() => {
          if (currentChat) {
            currentChat.closeMessageControl();
          } else {
            navigation.setParams({
              currentChat: null,
              selectedMessages: []
            });
          }
        }}
      />
    </View>
  );
};
const _menuControlHeader = navigation => {
  const currentChat = navigation.getParam('currentChat', null);
  const syncManager = SendBirdSyncManager.getInstance();
  const channel = navigation.getParam('channel', null);
  const selectedMessages = navigation.getParam('selectedMessages', []);
  return (
    <View style={style.headerRightContainer}>
      {selectedMessages.length === 1 &&
        selectedMessages[0].isUserMessage() &&
        selectedMessages[0].sender.userId === syncManager.currentUserId &&
        selectedMessages[0].messageId !== 0 && (
          <Button
            containerStyle={style.headerRightItemContainer}
            buttonStyle={style.headerRightItem}
            icon={<Icon name="md-create" color="#7d62d9" size={28} />}
            onPress={() => {
              if (currentChat) {
                currentChat.closeMessageControl();
              } else {
                navigation.setParams({
                  currentChat: null,
                  selectedMessages: []
                });
              }
              currentChat.switchToEditMode(selectedMessages[0]);
            }}
          />
        )}
      {selectedMessages.length === 1 && selectedMessages[0].isUserMessage() && (
        <Button
          containerStyle={style.headerRightItemContainer}
          buttonStyle={style.headerRightItem}
          icon={<Icon name="md-copy" color="#7d62d9" size={28} />}
          onPress={() => {
            if (selectedMessages.length > 0) {
              const message = selectedMessages[0];
              if (message.isUserMessage()) {
                Clipboard.setString(message.message);
                toast('The message is copied to Clipboard.');
                if (currentChat) {
                  currentChat.closeMessageControl();
                } else {
                  navigation.setParams({
                    currentChat: null,
                    selectedMessages: []
                  });
                }
              }
            }
          }}
        />
      )}
      {!selectedMessages[0].isAdminMessage() && selectedMessages[0].sender.userId === syncManager.currentUserId && (
        <Button
          containerStyle={style.headerRightItemContainer}
          buttonStyle={style.headerRightItem}
          icon={<Icon name="md-trash" color="#7d62d9" size={28} />}
          onPress={() => {
            Alert.alert('Delete', 'Do you want to delete this message?', [
              { text: 'Cancel' },
              {
                text: 'OK',
                onPress: () => {
                  for (let i = 0; i < selectedMessages.length; i++) {
                    if (selectedMessages[i].messageId === 0 && selectedMessages[i].requestState !== 'succeeded') {
                      componentContext.collection.deleteMessage(selectedMessages[i]);
                    } else {
                      channel.deleteMessage(selectedMessages[i], (res, err) => {
                        if (err) {
                          console.error(err);
                        }
                      });
                    }
                  }
                  navigation.setParams({
                    currentChat: null,
                    selectedMessages: []
                  });
                }
              }
            ]);
          }}
        />
      )}
    </View>
  );
};

export class ChatController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    const selectedMessages = navigation.getParam('selectedMessages', []);
    const isMenuControl = selectedMessages.length > 0;
    return {
      title: isMenuControl ? 'Message Selected' : `${createChannelTitle(channel)} (${channel.memberCount})`,
      headerLeft: isMenuControl ? _menuCloseHeader(navigation) : _chatCloseHeader(navigation),
      headerRight: isMenuControl ? _menuControlHeader(navigation) : _chatControlHeader(navigation)
    };
  };

  static getDerivedStateFromProps(props, state) {
    state.messages = [...props.messages];
    return state;
  }

  constructor(props) {
    super(props);
    this.collection = null;
    this.channel = this.props.navigation.getParam('channel', null);
    this.messageInput = null;
    this.state = {
      typedText: '',
      editedMessage: null,
      inputMode: 'chat',
      messages: []
    };
    componentContext = this;
  }

  _messageEventHandler(messages, action, reason) {
    switch (action) {
      case 'insert': {
        this.props.addMessages(messages);
        break;
      }
      case 'update': {
        if (reason === SendBirdSyncManager.MessageCollection.FailedMessageEventActionReason.UPDATE_RESEND_FAILED) {
          this.props.updateMessages(messages, true);
        } else {
          this.props.updateMessages(messages);
        }
        break;
      }
      case 'remove': {
        this.props.removeMessages(messages);
        break;
      }
      case 'clear': {
        this.props.clearMessages();
        break;
      }
    }
  }

  componentDidMount() {
    const sb = SendBird.getInstance();
    const channelHandler = new sb.ChannelHandler();
    channelHandler.onChannelChanged = channel => {
      this.updateChannel(channel);
    };
    channelHandler.onUserJoined = (channel, user) => {
      this.updateChannel(channel);
    };
    channelHandler.onUserLeft = (channel, user) => {
      if (sb.currentUser.userId === user.userId) {
        this.props.navigation.replace('NotFound', { channelUrl: channel.url });
      } else {
        this.updateChannel(channel);
      }
    };
    sb.addChannelHandler(`chat-${this.channel.url}`, channelHandler);

    this.collection = new SendBirdSyncManager.MessageCollection(this.channel);
    this.collection.limit = 50;

    const collectionHandler = new SendBirdSyncManager.MessageCollection.CollectionHandler();
    collectionHandler.onSucceededMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    collectionHandler.onFailedMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    collectionHandler.onPendingMessageEvent = (messages, action, reason) => {
      this._messageEventHandler(messages, action, reason);
    };
    this.collection.setCollectionHandler(collectionHandler);
    this.channel.markAsRead();

    SendBirdSyncManager.loggerLevel = 98765;
    this.collection.fetchSucceededMessages('prev', err => {
      this.collection.fetchSucceededMessages('next', err => {
        this.collection.fetchFailedMessages(err => {});
      });
    });

    navigator.push(this.channel.url);
    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
  }

  componentWillUnmount() {
    componentContext = null;
    this.props.clearMessages();
    if (this.collection) {
      this.collection.remove();
      this.collection = null;
    }
    navigator.pop();
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
  }

  handleBackPress = () => {
    const selectedMessages = this.props.navigation.getParam('selectedMessages', []);
    if (selectedMessages.length > 0) {
      this.closeMessageControl();
      return true;
    } else if (this.state.inputMode === 'edit') {
      this.switchToChatMode();
      return true;
    }
    return false;
  };

  updateChannel(channel) {
    this.channel = channel;
    this.props.navigation.setParams({
      title: `${createChannelTitle(channel)} (${channel.memberCount})`
    });
  }

  loadPrevious() {
    this.collection.fetchSucceededMessages('prev', err => {
      if (err) {
        console.log(err);
      }
    });
  }

  sendMessage() {
    if (this.state.typedText) {
      const pendingMessage = this.channel.sendUserMessage(this.state.typedText, (message, err) => {
        // To preserve order of pending messages being processed before same failed ones.
        setTimeout(() => {
          if (this.collection) {
            this.collection.handleSendMessageResponse(err, message);
          }
        }, 100);
      });
      if (this.collection) {
        this.collection.appendMessage(pendingMessage);
        this.setState({ typedText: '' });
      }
    }
  }

  updateMessage() {
    if (this.state.editedMessage && this.state.typedText) {
      const message = this.state.editedMessage;
      if (message.isUserMessage() && message.message !== this.state.typedText) {
        const messageContent = this.state.typedText;
        this.channel.updateUserMessage(
          message.messageId,
          messageContent,
          message.data,
          message.customType,
          (updatedMessage, err) => {
            if (!err) {
              this.collection.updateMessage(updatedMessage);
              this.switchToChatMode();
              Keyboard.dismiss();
            } else {
              toast('Failed to update the message.');
            }
          }
        );
      } else {
        this.switchToChatMode();
        Keyboard.dismiss();
      }
    } else {
      toast('Cannot update the message with empty content.');
    }
  }

  sendImage(source, options = {}) {
    setTimeout(() => {
      const tempMessage = this.channel.sendFileMessage(source, '', '', options.thumbnailSizes, (message, err) => {
        if (!err) {
          if (message && this.collection) {
            this.collection.appendMessage(message);
          }
        } else {
          toast('Failed to send image. ' + err.message);
          if (tempMessage && this.collection) {
            this.collection.deleteMessage(tempMessage);
          }
        }
      });
      if (this.collection) {
        this.collection.appendMessage(tempMessage);
      }
      this.setState({ typedText: '' });
    }, PAGE_RETURN_DELAY);
  }

  pickImage() {
    Permissions.check('photo').then(response => {
      if (response === 'authorized') {
        ImagePicker.showImagePicker(
          {
            title: 'Select Image',
            mediaType: 'photo',
            noData: true
          },
          response => {
            if (!response.didCancel && !response.error && !response.customButton) {
              const source = { uri: response.uri };
              if (response.name) {
                source['name'] = response.fileName;
              } else {
                paths = response.uri.split('/');
                source['name'] = paths[paths.length - 1];
              }
              if (response.type) {
                source['type'] = response.type;
              } else {
                /** For react-native-image-picker library doesn't return type in iOS,
                 *  it is necessary to force the type to be an image/jpeg (or whatever you're intended to be).
                 */
                if (Platform.OS === 'ios') {
                  source['type'] = 'image/jpeg';
                }
              }
              const thumbnailSizes = [{ maxWidth: 200, maxHeight: 200 }];
              this.sendImage(source, { thumbnailSizes });
            }
          }
        );
      } else if (response === 'undetermined') {
        Permissions.request('photo').then(response => {
          if (response === 'authorized') {
            this.pickImage();
          }
        });
      } else {
        toast('Permission denied. You declined the permission to access your photo.');
      }
    });
  }

  showMessageControl(message) {
    const syncManager = SendBirdSyncManager.getInstance();
    if (message.isUserMessage() || (message.isFileMessage() && message.sender.userId === syncManager.currentUserId)) {
      this.props.selectMessage(message);
      this.props.navigation.setParams({
        currentChat: this,
        selectedMessages: [message]
      });
    }
  }

  resendFailedMessage(message) {
    const pendingMessage = this.channel.resendUserMessage(message, (message, err) => {
      if (this.collection) {
        this.collection.handleSendMessageResponse(err, message);
      }
    });
  }

  closeMessageControl() {
    this.props.deselectAllMessages();
    this.props.navigation.setParams({
      currentChat: null,
      selectedMessages: []
    });
  }

  switchToChatMode() {
    this.setState({
      inputMode: 'chat',
      editedMessage: null,
      typedText: ''
    });
  }

  switchToEditMode(message) {
    this.setState({
      inputMode: 'edit',
      editedMessage: message,
      typedText: message.message
    });
    this.messageInput.focus();
  }

  render() {
    const mode = {
      chat: {
        left: {
          icon: 'md-cloud-upload',
          onPress: this.pickImage.bind(this)
        },
        right: {
          icon: 'md-send',
          onPress: this.sendMessage.bind(this)
        }
      },
      edit: {
        left: {
          icon: 'md-close',
          onPress: this.switchToChatMode.bind(this)
        },
        right: {
          icon: 'md-checkmark',
          onPress: this.updateMessage.bind(this)
        }
      }
    };
    const currentMode = mode[this.state.inputMode];
    return (
      <View style={style.container}>
        <View style={style.messageListViewContainer}>
          <FlatList
            data={this.state.messages}
            keyExtractor={message => (message.messageId > 0 ? message.messageId : message.reqId) + ''}
            onEndReached={this.loadPrevious.bind(this)}
            renderItem={bundle => (
              <MessageView
                channel={this.channel}
                message={bundle.item}
                onLongPress={this.showMessageControl.bind(this)}
                resendFailedMessage={this.resendFailedMessage.bind(this)}
              />
            )}
          />
        </View>
        <View style={style.messageInputContainer}>
          <Button
            containerStyle={style.uploadContainer}
            buttonStyle={style.upload}
            icon={<Icon name={currentMode.left.icon} size={28} color="#333" />}
            onPress={currentMode.left.onPress}
          />
          <Input
            placeholder="Your message"
            ref={input => {
              this.messageInput = input;
            }}
            containerStyle={style.formContainer}
            inputContainerStyle={style.formInputContainer}
            inputStyle={style.formInput}
            multiline={true}
            numberOfLines={3}
            autoCapitalize="none"
            autoCorrect={false}
            selectionColor="#ddd"
            underlineColorAndroid="transparent"
            value={this.state.typedText}
            onChangeText={typedText => this.setState({ typedText })}
          />
          <Button
            containerStyle={style.sendContainer}
            buttonStyle={style.send}
            icon={
              <Icon
                name={currentMode.right.icon}
                size={28}
                color={this.state.typedText.length > 0 ? '#7d62d9' : '#494e57'}
              />
            }
            onPress={currentMode.right.onPress}
          />
        </View>
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { chat } = previousState;
    return { ...chat };
  },
  Action
)(ChatController);
