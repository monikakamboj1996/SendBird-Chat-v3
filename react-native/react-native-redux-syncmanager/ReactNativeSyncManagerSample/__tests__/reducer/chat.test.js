import { reducer } from '../../src/reducer/chat';
import Action from '../../src/action/chat';

describe('reducer/chat', () => {
  test('checking initial state', () => {
    const action = { type: 'nonexistent' };
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  test('MESSAGE_SELECT', () => {
    const action = {
      type: Action.Type.MESSAGE_SELECT,
      payload: {
        message: {
          messageId: 1
        }
      }
    };
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 1
        },
        {
          messageId: 2
        },
        {
          messageId: 3
        },
        {
          messageId: 4
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('MESSAGE_DESELECT_ALL', () => {
    const action = {
      type: Action.Type.MESSAGE_DESELECT_ALL
    };
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 1,
          _selected: true
        },
        {
          messageId: 2,
          _selected: true
        },
        {
          messageId: 3
        },
        {
          messageId: 4
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('MESSAGE_ADD', () => {
    const action = {
      type: Action.Type.MESSAGE_ADD,
      payload: {
        messages: [
          {
            messageId: 5,
            createdAt: 5,
            sender: {
              userId: 4
            }
          },
          {
            messageId: 6,
            createdAt: 6,
            sender: {
              userId: 4
            }
          },
          {
            messageId: 7,
            createdAt: 7,
            sender: {
              userId: 5
            }
          }
        ]
      }
    };
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('MESSAGE_UPDATE', () => {
    const action = {
      type: Action.Type.MESSAGE_UPDATE,
      payload: {
        messages: [
          {
            messageId: 4,
            createdAt: 4,
            sender: {
              userId: 4
            },
            message: 'text 4'
          },
          {
            messageId: 3,
            createdAt: 3,
            sender: {
              userId: 4
            },
            message: 'text 3'
          }
        ]
      }
    };
    action.payload.messages.forEach(m => {
      m.isIdentical = om => {
        return om.messageId === m.messageId;
      };
    });
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          },
          message: 'old text 4',
          _isPreviousMessageSentBySameUser: true
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          },
          message: 'old text 3'
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('MESSAGE_REMOVE', () => {
    const action = {
      type: Action.Type.MESSAGE_REMOVE,
      payload: {
        messages: [
          {
            messageId: 3,
            createdAt: 3,
            sender: {
              userId: 4
            }
          }
        ]
      }
    };
    action.payload.messages.forEach(m => {
      m.isIdentical = om => {
        return om.messageId === m.messageId;
      };
    });
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('MESSAGE_CLEAR', () => {
    const action = {
      type: Action.Type.MESSAGE_CLEAR
    };
    const initState = {
      channel: null,
      messages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      failedMessages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('FAILED_MESSAGE_ADD', () => {
    const action = {
      type: Action.Type.FAILED_MESSAGE_ADD,
      payload: {
        messages: [
          {
            messageId: 5,
            createdAt: 5,
            sender: {
              userId: 4
            }
          },
          {
            messageId: 6,
            createdAt: 6,
            sender: {
              userId: 4
            }
          },
          {
            messageId: 7,
            createdAt: 7,
            sender: {
              userId: 5
            }
          }
        ]
      }
    };
    const initState = {
      channel: null,
      failedMessages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      messages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('FAILED_MESSAGE_REMOVE', () => {
    const action = {
      type: Action.Type.FAILED_MESSAGE_REMOVE,
      payload: {
        messages: [
          {
            messageId: 3,
            createdAt: 3,
            sender: {
              userId: 4
            }
          }
        ]
      }
    };
    action.payload.messages.forEach(m => {
      m.isIdentical = om => {
        return om.messageId === m.messageId;
      };
    });
    const initState = {
      channel: null,
      failedMessages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      messages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });

  test('FAILED_MESSAGE_CLEAR', () => {
    const action = {
      type: Action.Type.FAILED_MESSAGE_CLEAR
    };
    const initState = {
      channel: null,
      failedMessages: [
        {
          messageId: 4,
          createdAt: 4,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 3,
          createdAt: 3,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 2,
          createdAt: 2,
          sender: {
            userId: 4
          }
        },
        {
          messageId: 1,
          createdAt: 1,
          sender: {
            userId: 4
          }
        }
      ],
      inputMode: 'chat',
      messages: []
    };
    expect(reducer(initState, action)).toMatchSnapshot();
  });
});
