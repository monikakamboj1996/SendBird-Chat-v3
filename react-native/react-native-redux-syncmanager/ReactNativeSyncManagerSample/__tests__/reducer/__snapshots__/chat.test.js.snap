// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer/chat FAILED_MESSAGE_ADD 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [
    Object {
      "createdAt": 4,
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 3,
      "messageId": 3,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
  "inputMode": "chat",
  "messages": Array [],
}
`;

exports[`reducer/chat FAILED_MESSAGE_CLEAR 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [
    Object {
      "createdAt": 4,
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 3,
      "messageId": 3,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
  "inputMode": "chat",
  "messages": Array [],
}
`;

exports[`reducer/chat FAILED_MESSAGE_REMOVE 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [
    Object {
      "createdAt": 4,
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 3,
      "messageId": 3,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
  "inputMode": "chat",
  "messages": Array [],
}
`;

exports[`reducer/chat MESSAGE_ADD 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [
    Object {
      "_isPreviousMessageSentBySameUser": false,
      "createdAt": 7,
      "messageId": 7,
      "sender": Object {
        "userId": 5,
      },
    },
    Object {
      "_isPreviousMessageSentBySameUser": true,
      "createdAt": 6,
      "messageId": 6,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "_isPreviousMessageSentBySameUser": true,
      "createdAt": 5,
      "messageId": 5,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 4,
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 3,
      "messageId": 3,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
}
`;

exports[`reducer/chat MESSAGE_CLEAR 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [],
}
`;

exports[`reducer/chat MESSAGE_DESELECT_ALL 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [
    Object {
      "_selected": false,
      "messageId": 1,
    },
    Object {
      "_selected": false,
      "messageId": 2,
    },
    Object {
      "_selected": false,
      "messageId": 3,
    },
    Object {
      "_selected": false,
      "messageId": 4,
    },
  ],
}
`;

exports[`reducer/chat MESSAGE_REMOVE 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [
    Object {
      "createdAt": 4,
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "_isPreviousMessageSentBySameUser": true,
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
}
`;

exports[`reducer/chat MESSAGE_SELECT 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [
    Object {
      "_selected": true,
      "messageId": 1,
    },
    Object {
      "_selected": true,
      "messageId": 2,
    },
    Object {
      "_selected": true,
      "messageId": 3,
    },
    Object {
      "_selected": true,
      "messageId": 4,
    },
  ],
}
`;

exports[`reducer/chat MESSAGE_UPDATE 1`] = `
Object {
  "channel": null,
  "failedMessages": Array [],
  "inputMode": "chat",
  "messages": Array [
    Object {
      "_isPreviousMessageSentBySameUser": true,
      "createdAt": 4,
      "isIdentical": [Function],
      "message": "text 4",
      "messageId": 4,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "_isPreviousMessageSentBySameUser": undefined,
      "createdAt": 3,
      "isIdentical": [Function],
      "message": "text 3",
      "messageId": 3,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 2,
      "messageId": 2,
      "sender": Object {
        "userId": 4,
      },
    },
    Object {
      "createdAt": 1,
      "messageId": 1,
      "sender": Object {
        "userId": 4,
      },
    },
  ],
}
`;

exports[`reducer/chat checking initial state 1`] = `
Object {
  "channel": null,
  "inputMode": "chat",
  "messages": Array [],
}
`;
